### ISSUE TITLE
Recognise Antora structure in order to support images + examples

### ISSUE THREAD

**Issue by danhaywood** [asciidoctor/asciidoctor-intellij-plugin#373]
Initial support for Antora would be great.

As per Dan's comment in #330, if the plugin recognises an Antora structure, then it could switch into some sort of "Antora" mode.

In this mode would mean support for recognising Antora's conventions for include:, image: and xref:

* image family for image, eg
    * image:foo/some-image.png[]
* example family for includes, eg
    * include::example$Foo.java[]
* xrefs to other modules:
    * xref:OTHER_COMPONENT:OTHER_MODULE:index.adoc[]

**Comment by danhaywood** [comment]
I currently simulate the first by including this _attributes.adoc[] in every file:

include::_attributes.adoc[]

where the "local" _attributes.adoc itself looks something like:

ifndef::env-site,env-github[]
:moduledir: ../..
include::{moduledir}/_attributes.adoc[]
endif::[]

which is only activated in the IDE, not by Antora.  

The :moduledir attribute then depends on the depth of the folder hierarchy, and then links to a module level _attributes.adoc :

ifndef::env-site,env-github[]
:attachmentsdir: {moduledir}/attachments
:examplesdir: {moduledir}/examples
:imagesdir: {moduledir}/images
:partialsdir: {moduledir}/partials
endif::[]

It's the setting of "imagesdir" that works here

**Comment by mojavelinux** [comment]
Recall that the Antora structure can be detected by the presence of antora.yml.

Creating references within the module is rather straightforward since all Antora component versions follow the same structure.

The trickier part is references to other components and component versions. The plugin won't necessarily know about those since they may not even be on the same filesystem. In the future, I anticipate being able to download a manifest from a previous Antora run that can provide a reverse lookup. This is similar to how references are handled in Java to other libraries. I'm happy to discuss the design of such a system. I don't expect the plugin to always have to scan for this information, though it certain could make use of whatever information is available.

**Comment by ahus1** [comment]
Hi @danhaywood - an Antora would be a great thing. 

While this is not implemented, some existing features might help your: 
* conditional rendering in the IntelliJ plugin: use `ifndef::env-idea[]`, see [Conditional Rendering](https://github.com/asciidoctor/asciidoctor-intellij-plugin/wiki/Conditional-Rendering) wiki entry for details
* Support of an `.asciidoctorconfig` file might save you the include in every file and also the conditional settings as well as it will only work in the IntelliJ plugin preview anyway. See [Support project specific configurations](https://github.com/asciidoctor/asciidoctor-intellij-plugin/wiki/Support-project-specific-configurations) for details. It also adds an attribute `{asciidoctorconfigdir}` that's filled with the absolute path of the folder where the configuration file resides. 

This will not take you 100% of the way, but might ease one or two symptoms. Please use it only for quick-wins as Antora support at least with local intra-module references is on the agenda (but my spare time is limited, and it might take some months).

**Comment by mojavelinux** [comment]
> it might take some months

We definitely want to take our time with this one, so I actually consider that a good thing. I'm open to brainstorming ideas as we think about how this will work best, which could save time implementing it in the long run. And if we need changes in Antora, that gives me time to make them.

**Comment by ahus1** [comment]
I've published a pre-release 0.30.23 that takes a first small step: it pre-propulates `{partialsdir}` according to the Antora structure with the absolute path to either the `partials` or the `pages/_partials` pages of the local's module.

I gave it a try with the fedora quick docs folder: https://pagure.io/fedora-docs/quick-docs.git
It works for the editor and the preview.

I could do the same for the other attributes `attachmentsdir` and `examplesdir`, but I assume at least for `imagesdir` it needs to a relative path?

![image](https://user-images.githubusercontent.com/3957921/71447710-4eac8a80-2732-11ea-9d14-6a8e258e7b6c.png)

**Comment by ahus1** [comment]
The latest pre-release 0.30.24 now supports all four attributes: attachmentsdir, examplesdir, imagesdir, partialsdir. This should allow you to skip `_attributes.adoc` in every file.

The plugin also supports xref for module-local references.

As this is a first a first, experimental support, I might have missed corner cases or added bugs. Please comment with examples so I can fix them.

The pre-release of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/CONTRIBUTING.adoc#eap).

**Comment by djencks** [comment]
As an alternative approach, that would be more extensible to other components and versions, one could either modify the html5 processor, as Antora does, to detect and interpret references, or perhaps add a tree processor extension before the html5 rendering step to detect and interpret references.

**Comment by mojavelinux** [comment]
@ahus1 This is a good first step, but note that the use of partialsdir and examplesdir is now deprecated in Antora. The preferred approach is to use the universal cross referencing syntax. For example:

```
include::partial$unreviewed-message.adoc[]
```

Images don't need this prefix since the family is implicit. So that path (at least within the module) is relative to the images folder.

As @djencks points out, the right way to solve this is to intercept the references and resolve them directly (rather than indirectly using attributes).

> add a tree processor extension before the html5 rendering step to detect and interpret references.

@djencks This plugin doesn't need to use Asciidoctor extensions since its parsing the AsciiDoc directly, at least in the editor part.

**Comment by djencks** [comment]
@mojavelinux The editor wouldn't use asciidoctor extensions, but the renderer would... I was thinking of looking into that part since I'm more familiar with it.

**Comment by mojavelinux** [comment]
Yep. So there are really two separate concerns here. One is rendering in the preview pane and the other is editor assistance. We need to make sure we're clear which one we're talking about as we work through the design for Antora support (speaking to myself as well).

**Comment by ahus1** [comment]
@djencks - as images already work in the preview, the next thing for the preview would be includes with the universal cross reference syntax with a $. AFAIK this can only be solved as an Asciidoctor extension. You are free to use all types of extensions. For now it would be ok the make things work inside a local Antora module. 

For xrefs and links to work there's also the option to add this to the link handler in the IDE. It might have more options to resolve to external modules, therefore I would consider this a lower priority here.

```
// include content into a page
include::partial$some-recurring-content.adoc[]

// include snippets from examples
[source,java]
----
include::example$test.java[tags=e1]
----

// offer downloads
link:attachment$file.zip[Download]
```

The HTML creation is bundled in [AsciiDoc.java]( https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/java/org/asciidoc/intellij/AsciiDoc.java), and there are Unit tests in AsciiDocTest.java. To make the magic around Antora detection work you'll need to point to an Antora folder structure. Please add one to suit your needs below /testData.

For handling links in the preview there is JavaFxHtmlPanel#openLink() - this does all the magic to open a file in the editor when you click on a link in the preview. This will need some additions to support the xref/link syntax.

**Comment by ahus1** [comment]
@djencks - I've added first support for module-local Antora families in the editor.

I then gave it a go how to support Antora families for includes without touching Ruby. I came up with a solution of an Java based IncludeProcessor, see e3febfdb060bf00c8f918a7605b6f8a91146365e for the commit. It's still quite crude (for example I map the `example$` family prefix to the matching `{examplesdir}` attribute. Still it shows how I could translate Antora includes to standard AsciiDoc includes.

I think that by adding some more lines of Java code and injecting some more context the I'll be to support Antora cross-modules references as well. I'll provide another pre-release for this.

**Comment by ahus1** [comment]
There's a new pre-release 0.30.27 that is still restricted to module-local references, but now support families for includes in editor and preview / cc: @danhaywood
Over the next days I'll give it a test run myself and will add some more tests. Anyone who wants to give it a try is welcome and provide feedback here. 

Next-up is cross-module support.

The new pre-release of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/CONTRIBUTING.adoc#eap).

**Comment by danhaywood** [comment]
HI Alexander,
Just installed the pre-release ... this is great work!  I can confirm that
I've been able to remove all of my hacky _attributes.adoc files and that
the example$ family resolution works fine.

Looking forward to cross-module support as well.  I guess to support that
you'll need to detect site.yml in order to locate the roots of each of the
modules.  As Dan Allen points out, site.yml does support remote URLs,
supporting that in the IDE is non-trivial without some additional infra
from Antora itself.

As a more achievable goal, could I suggest that when you parse site.yml,
for now you ignore any remote URLs and just handle local ones.  For
example, in the open source project I work on, we maintain a "local"
site.yml for authoring purposes, see [1]  As you can see, all of the URLs
here are local so all the info required for an IDE is at hand.

Oh, and of course you are welcome to clone that repo if you want for your
testing of cross-module support.  The generated website can be found here
[2] with the main content under "docs", [3]

Cheers
Dan

[1] https://github.com/apache/isis/blob/master/site.yml
[2] https://apache-isis-committers.github.io/isis-nightly/
[3] https://apache-isis-committers.github.io/isis-nightly/toc/about.html

**Comment by ahus1** [comment]
Hi Dan (Haywood),

I see that there are lots of components with one antora.yml each, for example refguide, testing, etc. Can you point me to one of the modules that you think has the progressed most regarding content from your point of view? I could take it as something to test with. 

I see that sometimes you use the component's own name in includes or xrefs, for example `xref:security:ROOT:usage-by-isis-viewers.adoc[Usage by Isis Viewers]` is part of the `security` component. AFAIK you could leave of `security:` and it would still work. 

BTW: instead of looking for a site.yml I might look for the different antora.yml files and look for the name specified inside the file. Let's see how that works out.

**Comment by ahus1** [comment]
(and thanks for giving it a try, it's good to know that the plugin is on the right track, although we're not there yet)

**Comment by djencks** [comment]
Be sure to understand the difference between source location (url + start_path), component, and module.  Do you have full support for a single component in one source location working, or just one module in a single component in one source location?

The preferred name for the playbook is now antora-playbook.yml rather than site.yml, although it can be named anything.

I’m not sure but Dan may have been using “module” instead of “component”.

The playbook can also contribute asciidoctor attributes to the rendering, along with specifying the source locations.

Theoretically, once a project is built locally on a particular machine by antora, editors could use the antora cache to get the remote content from the bare git repo in the cache.  I don’t know if this would be easier or harder than making one’s own clone.

You might find some of the projects at https://gitlab.com/djencks/simple-examples <https://gitlab.com/djencks/simple-examples> useful for testing.  For instance multiple-components/xrefs has all possible choices of xref syntax.

Hope this helps,
David Jencks

**Comment by danhaywood** [comment]
HI Alexander,

You could use site-security.xml .

As you can see there are quite a few site-xxx.yml files, each of which is a
subset of the main site.yml .  If you generate from site-security.yml then
of course there will be some invalid links.

So far as site-security.yml goes, it does have multiple components ("toc"
and "security") and there are multiple modules in the security component
(distributed across multiple source roots, which is a neat trick I learnt
along the way).

Cheers

**Comment by danhaywood** [comment]
@David ... Yep, when I said "cross-module" I was conflating both components
and modules, apologies.  I guess I meant xref to modules that could be in
any component.

@Alexandar - probably David's example projects would be better (faster) to
test with, but it'd be great if you could also use mine for a bit of "real
world" testing also.

Cheers

**Comment by ahus1** [comment]
I've made good progress supporting Antora components and modules for include and xref in both editor and preview. 

If an xref/includes leaves the current component or module, the plugin will search all antora.yml files in the project to find a match. The plugin doesn't build an index on that yet, for now it still seems to be fast enough to support the 75 antora.yml files in the isis project.

The plugin supports the isis approach to have multiple antora.yml files with the same component name, but different modules. When there are Antoras with the same component name and the same module, it will choose the once closest in the tree to resolve the reference.

Using that approach it doesn't rely on an Antora playbook to discover references. 

Expect a new pre-release soon.

**Comment by danhaywood** [comment]
Great stuff, Alexander. Looking forward to trying out that next preview
build!

**Comment by ahus1** [comment]
A new pre-release 0.30.28 is available that supports cross-component and cross-modules references for xref and include.

The linking in the editor works for Antora-prefixes, file, directories and anchors. On windows use Ctrl-click and Ctrl-B. The auto-complete for files, directories and anchors work. There is no auto-complete for Antora-prefixes yet. Clicking on links in the preview doesn't work for Antora references yet.

The auto-complete for anchors (used extensively in Antora's own documentation) relies either block IDs to be set, or for the idprefix/idseparator attributes to be configured in an .asciidoctorconfig file in order to generate the proper section headings automatically. Use Ctrl-Space to trigger autocomplete, and use tab to select an entry that should replace the current entry.

I've given it a test run on the Isis project and Antora's own docs do work on corner cases. Please report corner cases that I missed here. I've pushed some changes to the Isis project here: https://github.com/apache/isis/pull/170 (already merged while I was writing this),

Final disclaimer: this is still an early preview. Due to time constraints this will be the last preview for this week. Enjoy the new pre-release, and the short movies below.

![xref-include-antora-cross-component](https://user-images.githubusercontent.com/3957921/71919998-343ccd00-3186-11ea-8446-e2fea1bf80ee.gif)

![xref-autocomplete](https://user-images.githubusercontent.com/3957921/71920246-aad9ca80-3186-11ea-9279-97784d5f133e.gif)

**Comment by danhaywood** [comment]
Many thanks for this Alexander ; and the doc improvements for isis - a
welcome bonus!

I've installed the new version, seems to works nicely as the gif's indicate.

One suggestion for the next time you get a chance to work on this ... it
would be fabulous if there was intellisense for the component and then
module parts of the xref: or include:: macros.

In other words, if I type:

           xref: <ctrl+enter>

then I would expect to be presented with a list consisting of all of:

* all the available components
* all the available modules in the current component
* all of the available pages in the current module.

At the moment the intellisense offers only the last of these.

And similarly, if I've entered the component, eg:

        xref:extensions:  <ctrl+eter

then I would be presented with a list of all of the modules in that
component (and possible a list of pages if "extensions" was also a module
in the current component).

I have one further enhancement idea also, which is to automatically look up
the title of the referenced page for the "[...]".  So, if
extensions:ROOT:about.adoc has the title "Extensions Catalog", then typing:

       xref:extensions:ROOT:[  <ctrl+shift+enter>

should lookup the title and autocomplete.

I suspect you have these enhancement ideas planned already, but anyway,
that's my 2c.


Thx again
Dan

**Comment by djencks** [comment]
Would it make sense to also have intellisense for the component version?  I’d expect that choosing the version after the component would be best, although it occurs to the left in the xref.  I haven’t used intellisense for perhaps 8 years so maybe this doesn’t make sense …

David Jencks

**Comment by danhaywood** [comment]
Absolutely, that *does* make sense.    I guess the docs at [1] make it
pretty clear how this might work.


[1] https://docs.antora.org/antora/2.2/page/page-id/#structure

**Comment by ahus1** [comment]
@danhaywood  / @djencks  - For autocompletion of component and module I will have a go as part of this ticket. Once this is complete (and no-one has reported problems with this new feature) I'll publish a public release a version of the plugin (no longer preview).

Looking up the title would be out of scope for this ticket, and probably a performance issue as each page needs to be parsed. This would make it slow, therefore I would vote against it. Feel free to open a ticket to discuss ins and outs.

Intellisense for the component version will also be out-of-scope for this ticket, as it would require to parse not only the current working directory's contents, but also branches. Please open a separate issue for this.

**Comment by danhaywood** [comment]
Sounds sensible to me. Intellisense of components and modules (no branch
versions) would be the most common use case imho.

Looking forward to seeing it/trying it!

**Comment by danhaywood** [comment]
I've created #377 for the branch@ extension to intellisense, referencing back to this ticket.

**Comment by danhaywood** [comment]
And #378 for xref title idea.

**Comment by djencks** [comment]
Well….
That will work sometimes.  If you are working on a prerelease version, leaving out the version in xrefs will point you to the latest version, not the prerelease version of the other component, which is quite possibly what you have checked out.

Also, you can have all the antora versions in the same git branch, or separate git repos or…. 

Would it be possible to insert the version specified in the component descriptor you get the component name from?  I think that would mean the component intellisense would show name+version rather than just name.

Great work anyway!!

**Comment by danhaywood** [comment]
@djencks  - perhaps you could add those clarifying comments onto #377 since Alexander has descoped this suggestion from this ticket, at least.

**Comment by djencks** [comment]
soon….

What happens now if there are two versions of the same component open in intellij?  I’d expect the component name shows up twice.  So for this issue, if it’s easy to show the  version with the component name, and include it in the xref, I think that would be useful.  Organizing the versions and components separately could wait.

**Comment by ahus1** [comment]
The pre-release 0.30.29 adds auto-completion for family and component/module. Component are always suggested in combination as shown below. 

![image](https://user-images.githubusercontent.com/3957921/72222515-a5ed9000-3565-11ea-8e4a-3a4f586d4f4f.png)

I wonder if there should be an explicit list of macros that support the Antora resource ID syntax, and those who don't. And those that support adding the family, those who have a default family, and those that must not have a family.

Next up: auto-complete component/module for images to choose images outside the current module and make them appear in the preview.

The new pre-release of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/CONTRIBUTING.adoc#eap).

**Comment by ahus1** [comment]
A new pre-release 0.30.30 is available that includes auto-complete component/module for images to choose images outside the current module and make them appear in the preview.

I consider this release feature-complete for a first public release. 

@danhaywood, @djencks - please give it a try; maybe I've missed a corner case or two. Please report bugs as comments in this issue. Please create new issues for feature that are currently missing that should be added in a future release.

**Comment by ahus1** [comment]
A new pre-release 0.30.31 restricts Antora intellisense to macros "image", "include", "xref" - I've also added "video", "audio", but I am not sure if they work with Antora.

This is also the first pre-release to support Antora syntax for image inline macros (one colon). Previously only image block macros worked (two colons).

This pre-release contains autocomplete for link attributes, see #383

The pre-release of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/CONTRIBUTING.adoc#eap).

**Comment by ahus1** [comment]
I've promoted this release to GA. The functionality asked for in the initial request is now implemented, therefore I close the ticket. 

Please report bugs and other Antora features as new GitHub issues is possible. 

Only if it is directly related to this ticket, please comment here and I'll re-open the ticket. 

I'll label this and follow-up tickets related to Antora with a label so it's easier to follow up: https://github.com/asciidoctor/asciidoctor-intellij-plugin/labels/Antora
